lib/ui/window.dart

```diff
@@ -286,31 +569,212 @@ class Locale {
     }
   }
 
+  // Unicode language variant codes.
+  //
+  // This iterable provides variants normalized to lowercase and in sorted
+  // order, as per the Unicode LDML specification.
+  //
+  // FIXME: decide whether this returns a single string or something fancier.
+  Iterable<String> get variants => _variants ?? const <String>[];
+
+  // The private _variants field must have variants in lowercase and already
+  // sorted: constructors must construct it as such.
+  final List<String> _variants;
+
+  // A map representing all Locale Identifier extensions.
+  //
+  // Keys in this ordered map must be sorted, and both keys and values must all
+  // be lowercase: constructors must construct it as such.
+  //
+  // This map simultaneously reprents T-extensions, U-extensions, other
+  // extensions and the private use extensions. Implementation detailsf:
+  //
+  // * The 't' entry represents the optional "tlang" identifier of the T
+  //   Extension. If the T Extension is present but has no tlang value, the 't'
+  //   map entry's value must be an empty string.
+  // * The 'u' entry represents the optional attributes of the U Extension. They
+  //   must be sorted in alphabetical order, separated by hyphens, and be
+  //   lowercase. If the U Extension is present but has no attributes, the 'u'
+  //   map entry's value must be an empty string.
+  // * U-Extension keyword keys, matching _reKey, and T-Extension fields in the
+  //   map, whos field separator subtags match _reTkey, are directly entered
+  //   into this map. (These regular expressions don't match the same keys.)
+  // * Other singletons are entered directly into the map, with all
+  //   values/attributes associated with that singleton as the map entry's
+  //   value.
+  final collection.LinkedHashMap<String, String> _extensions;
+
+  // Produces the Unicode BCP47 Locale Identifier for this locale.
+  //
+  // If the unnamed constructor was used with bad parameters, the result might
+  // not be standards-compliant.
+  //
+  // FIXME/TODO: this must not be submitted as a public function until we've
+  // made a final decision on toString() behaviour.
+  String toLanguageTag() {
+    final StringBuffer out = StringBuffer(languageCode);
+    if (scriptCode != null) {
+      out.write('-$scriptCode');
+    }
+    if (_countryCode != null && _countryCode != '') {
+      out.write('-$countryCode');
+    }
+    for (String v in variants) {
+      out.write('-$v');
+    }
+    if (_extensions != null && _extensions.isNotEmpty) {
+      out.write(_extensionsToString(_extensions));
```

Comment:
You are creating a string using concatenation (actually a string buffer), then adding that string to a string buffer.
Just pass the original string buffer to the helper method and have it add its parts directly.